File: 124.binary-tree-maximum-path-sum.js
/**
 * 这题不能用直接递归，
只能用 helper DFS function
因为 DFS 返回值 和 res 是无关联的
 */

var maxPathSum = function (root) {
  let res = -Infinity;
  const DFS = (node) => {
    if (!node) return 0;
    let left = Math.max(DFS(node.left), 0);
    let right = Math.max(DFS(node.right), 0);
    res = Math.max(res, left + right + node.val); // 记录一个完整的路径
    return Math.max(left, right) + node.val; // 构成当前的最大值 返回
  };
  DFS(root);
  return res;
};

File: solutions/124.binary-tree-maximum-path-sum.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_path_sum(root)
  res = -2e31
  dfs = lambda do |node|
    return 0 if node.nil?
    left_sum = max(0, dfs.call(node.left))
    right_sum = max(0, dfs.call(node.right))
    res = max(res, node.val + left_sum + right_sum)
    node.val + max(left_sum, right_sum)
  end

  dfs.call(root)
  res
end

def max(*nums) = nums.max
